File: python/agentbay/session_params.py
from typing import Dict, Optional, List, TYPE_CHECKING
from agentbay.context_sync import ContextSync, DeletePolicy, SyncPolicy, UploadPolicy, ExtractPolicy, RecyclePolicy, BWList, WhiteList
from agentbay.config import _BROWSER_FINGERPRINT_PERSIST_PATH
from agentbay.logger import get_logger
from agentbay.api.models._create_mcp_session_request import ExtraConfigs

# Initialize _logger for this module
_logger = get_logger("session_params")

if TYPE_CHECKING:
    from agentbay.extension import ExtensionOption
    from agentbay.browser.browser import BrowserFingerprintContext


class BrowserContext:
    """
    Browser context configuration for session.

    This class provides browser context configuration for cloud sessions, supports browser fingerprint
    context persistence, supports automatic extension synchronization when ExtensionOption is provided.

    Key Features:
    - Browser context binding for sessions
    - Automatic browser data upload on session end
    - Optional browser fingerprint integration with automatic context sync generation
    - Optional extension integration with automatic context sync generation
    - Clean API with ExtensionOption encapsulation

    Attributes:
        context_id (str): ID of the browser context to bind to the session
        auto_upload (bool): Whether to automatically upload browser data when session ends
        fingerprint_context (Optional[BrowserFingerprintContext]):
            Browser fingerprint context configuration object containing fingerprint_context_id.
        extension_option (Optional[ExtensionOption]): Extension configuration object containing
                                                     context_id and extension_ids.
        extension_context_id (Optional[str]): ID of the extension context for browser extensions.
                                             Set automatically from extension_option.
        extension_ids (Optional[List[str]]): List of extension IDs to synchronize.
                                           Set automatically from extension_option.
        extension_context_syncs (Optional[List[ContextSync]]): Auto-generated context syncs for extensions.
                                                              None if no extension configuration provided,
                                                              or List[ContextSync] if extensions are configured.

    Extension Configuration:
    - **ExtensionOption**: Pass an ExtensionOption object with context_id and extension_ids
    - **No Extensions**: Don't provide extension_option parameter (extension_context_syncs will be None)

    Usage Examples:
        ```python
        # With extensions using ExtensionOption
        from agentbay.extension import ExtensionOption

        ext_option = ExtensionOption(
            context_id="my_extensions",
            extension_ids=["ext1", "ext2"]
        )

        browser_context = BrowserContext(
            context_id="browser_session",
            auto_upload=True,
            extension_option=ext_option
        )

        # Without extensions (minimal configuration)
        browser_context = BrowserContext(
            context_id="browser_session",
            auto_upload=True
        )
        # extension_context_syncs will be None
        ```
    """

    def __init__(self, context_id: str, auto_upload: bool = True,
                 extension_option: Optional["ExtensionOption"] = None,
                 fingerprint_context: Optional["BrowserFingerprintContext"] = None):
        """
        Initialize BrowserContext with optional extension support.

        Args:
            context_id (str): ID of the browser context to bind to the session.
                             This identifies the browser instance for the session.

            auto_upload (bool, optional): Whether to automatically upload browser data
                                        when the session ends. Defaults to True.

            fingerprint_context (Optional[BrowserFingerprintContext], optional):
                        Browser fingerprint context configuration object containing fingerprint_context_id.
                        Defaults to None.

            extension_option (Optional[ExtensionOption], optional): Extension configuration object containing
                                                                   context_id and extension_ids. This encapsulates
                                                                   all extension-related configuration.
                                                                   Defaults to None.
        Fingerprint Configuration:
            - **BrowserFingerprintContext**: Use fingerprint_context parameter with a BrowserFingerprintContext object
            - **No Fingerprint**: Don't provide fingerprint_context parameter (fingerprint_context_sync will be None)

        Fingerprint Auto-generation:
            - fingerprint_context_sync is automatically generated when fingerprint_context is provided
            - fingerprint_context_sync will be None if no fingerprint_context is provided
            - fingerprint_context_sync will be a ContextSync if fingerprint_context is valid

        Extension Configuration:
            - **ExtensionOption**: Use extension_option parameter with an ExtensionOption object
            - **No Extensions**: Don't provide extension_option parameter

        Auto-generation:
            - extension_context_syncs is automatically generated when extension_option is provided
            - extension_context_syncs will be None if no extension_option is provided
            - extension_context_syncs will be a List[ContextSync] if extension_option is valid

        Examples:
            ```python
            # With extensions using ExtensionOption
            from agentbay.extension import ExtensionOption

            ext_option = ExtensionOption(
                context_id="my_extensions",
                extension_ids=["ext1", "ext2"]
            )

            browser_context = BrowserContext(
                context_id="browser_session",
                auto_upload=True,
                extension_option=ext_option
            )

            # Without extensions (minimal configuration)
            browser_context = BrowserContext(
                context_id="browser_session",
                auto_upload=True
            )
            # extension_context_syncs will be None
            ```
        """
        self.context_id = context_id
        self.auto_upload = auto_upload
        self.extension_option = extension_option
        self.fingerprint_context = fingerprint_context

        # Handle extension configuration from ExtensionOption
        if extension_option:
            # Extract extension information from ExtensionOption
            self.extension_context_id = extension_option.context_id
            self.extension_ids = extension_option.extension_ids
            # Auto-generate extension context syncs
            self.extension_context_syncs = self._create_extension_context_syncs()
        else:
            # No extension configuration provided
            self.extension_context_id = None
            self.extension_ids = []
            self.extension_context_syncs = None
        
        # Handle browser fingerprint configuration from BrowserFingerprintContext
        if fingerprint_context:
            self.fingerprint_context_id = fingerprint_context.fingerprint_context_id
            # Auto-generate fingerprint context sync
            self.fingerprint_context_sync = self._create_fingerprint_context_sync()
        else:
            self.fingerprint_context_id = None
            self.fingerprint_context_sync = None

    def _create_extension_context_syncs(self) -> List[ContextSync]:
        """
        Create ContextSync configurations for browser extensions.

        This method is called only when extension_option is provided and contains
        valid extension configuration (context_id and extension_ids).

        Returns:
            List[ContextSync]: List of context sync configurations for extensions.
                              Returns empty list if extension configuration is invalid.
        """
        if not self.extension_ids or not self.extension_context_id:
            return []

        # Create whitelist for each extension ID
        white_lists = [WhiteList(path=f"{ext_id}", exclude_paths=[]) for ext_id in self.extension_ids]

        # Create sync policy for extensions
        sync_policy = SyncPolicy(
            upload_policy=UploadPolicy(auto_upload=False),
            delete_policy=DeletePolicy(sync_local_file=False),
            extract_policy=ExtractPolicy(extract=True, delete_src_file=True),
            recycle_policy=RecyclePolicy.default(),
            bw_list=BWList(white_lists=white_lists)
        )

        # Create context sync for extensions
        extension_sync = ContextSync(
            context_id=self.extension_context_id,
            path="/tmp/extensions/",
            policy=sync_policy
        )

        return [extension_sync]
    
    def _create_fingerprint_context_sync(self) -> ContextSync:
        """
        Create ContextSync configurations for browser fingerprint.
        
        This method is called only when fingerprint_context is provided and contains
        valid fingerprint configuration (fingerprint_context_id).
        
        Returns:
            ContextSync: Context sync configurations for fingerprint.
                         Returns None if fingerprint configuration is invalid.
        """
        if not self.fingerprint_context_id or not self.fingerprint_context_id.strip():
            return None

        # Create whitelist for fingerprint context
        white_lists = []

        # Create sync policy for fingerprint
        sync_policy = SyncPolicy(
            upload_policy=UploadPolicy(auto_upload=False),
            extract_policy=ExtractPolicy(extract=True, delete_src_file=True),
            bw_list=BWList(white_lists=white_lists)
        )

        # Create context sync for fingerprint
        fingerprint_sync = ContextSync(
            context_id=self.fingerprint_context_id,
            path=_BROWSER_FINGERPRINT_PERSIST_PATH,
            policy=sync_policy
        )

        return fingerprint_sync


    def get_extension_context_syncs(self) -> List[ContextSync]:
        """
        Get context syncs for extensions.

        Returns:
            List[ContextSync]: Context sync configurations for extensions.
                               Returns empty list if no extensions configured.
        """
        return self.extension_context_syncs or []

    def get_fingerprint_context_sync(self) -> ContextSync:
        """
        Get context sync for fingerprint.

        Returns:
            ContextSync: Context sync configurations for fingerprint.
                         Returns None if fingerprint configuration is invalid.
        """
        return self.fingerprint_context_sync


class CreateSessionParams:
    """
    Parameters for creating a new session in the AgentBay cloud environment.

    Attributes:
        labels (Optional[Dict[str, str]]): Custom labels for the Session. These can be
            used for organizing and filtering sessions.
        context_syncs (Optional[List[ContextSync]]): List of context synchronization
            configurations that define how contexts should be synchronized and mounted.
        browser_context (Optional[BrowserContext]): Optional configuration for browser data synchronization.
        is_vpc (Optional[bool]): Whether to create a VPC-based session. Defaults to False.
        policy_id (Optional[str]): Policy id to apply when creating the session.
        enable_browser_replay (Optional[bool]): Whether to enable browser recording for the session. Defaults to False.
        extra_configs (Optional[ExtraConfigs]): Advanced configuration parameters for mobile environments.
        framework (Optional[str]): Framework name for tracking (e.g., "langchain"). Defaults to empty string (direct call).
    """

    def __init__(
        self,
        labels: Optional[Dict[str, str]] = None,
        image_id: Optional[str] = None,
        context_syncs: Optional[List[ContextSync]] = None,
        browser_context: Optional[BrowserContext] = None,
        is_vpc: Optional[bool] = None,
        policy_id: Optional[str] = None,
        enable_browser_replay: Optional[bool] = None,
        extra_configs: Optional[ExtraConfigs] = None,
        framework: Optional[str] = None,
    ):
        """
        Initialize CreateSessionParams.

        Args:
            labels (Optional[Dict[str, str]], optional): Custom labels for the Session.
                Defaults to None.
            image_id (Optional[str], optional): ID of the image to use for the session.
                Defaults to None.
            context_syncs (Optional[List[ContextSync]], optional): List of context
                synchronization configurations. Defaults to None.
            browser_context (Optional[BrowserContext], optional): Browser context configuration.
                If extension_ids are provided in BrowserContext, extension syncs will be
                automatically added. Defaults to None.
            is_vpc (Optional[bool], optional): Whether to create a VPC-based session.
                Defaults to False.
            policy_id (Optional[str], optional): Policy id to apply when creating the session.
                Defaults to None.
            enable_browser_replay (Optional[bool], optional): Whether to enable browser recording for the session.
                Defaults to False.
            extra_configs (Optional[ExtraConfigs], optional): Advanced configuration parameters for mobile environments.
                Defaults to None.
            framework (Optional[str], optional): Framework name for tracking (e.g., "langchain").
                Defaults to None (empty string in tracking, indicating direct call).
        """
        self.labels = labels or {}
        self.image_id = image_id

        # Start with provided context_syncs
        all_context_syncs = list(context_syncs or [])

        # Add extension context syncs from browser_context if available
        if browser_context and browser_context.extension_context_syncs:
            all_context_syncs.extend(browser_context.extension_context_syncs)
            _logger.info(f"Added {len(browser_context.extension_context_syncs)} extension context sync(s) from BrowserContext")

        # Add fingerprint context sync from browser_context if available
        if browser_context and browser_context.fingerprint_context_sync:
            all_context_syncs.append(browser_context.fingerprint_context_sync)
            _logger.info(f"Added fingerprint context sync from BrowserContext")

        self.context_syncs = all_context_syncs
        self.browser_context = browser_context
        self.is_vpc = is_vpc if is_vpc is not None else False
        self.policy_id = policy_id
        self.enable_browser_replay = enable_browser_replay if enable_browser_replay is not None else False
        self.extra_configs = extra_configs
        self.framework = framework or ""


class ListSessionParams:
    """
    Parameters for listing sessions with pagination support.

    Attributes:
        max_results (int): Number of results per page.
        next_token (str): Token for the next page.
        labels (Dict[str, str]): Labels to filter by.
    """

    def __init__(
        self,
        max_results: int = 10,
        next_token: str = "",
        labels: Optional[Dict[str, str]] = None,
    ):
        """
        Initialize ListSessionParams with default values.

        Args:
            max_results (int, optional): Number of results per page. Defaults to 10.
            next_token (str, optional): Token for the next page. Defaults to "".
            labels (Optional[Dict[str, str]], optional): Labels to filter by.
                Defaults to None.
        """
        self.max_results = max_results
        self.next_token = next_token
        self.labels = labels if labels is not None else {}

File: typescript/src/session-params.ts
import { ContextSync, SyncPolicy, newUploadPolicy, newExtractPolicy,newRecyclePolicy,Lifecycle,DownloadStrategy, WhiteList, newDeletePolicy } from "./context-sync";
import { ExtensionOption } from "./extension";
import { BrowserFingerprintContext } from "./browser";
import { ExtraConfigs, extraConfigsToJSON } from "./types/extra-configs";

// Browser fingerprint persistent path constant (moved from config.ts)
const BROWSER_FINGERPRINT_PERSIST_PATH = "/tmp/browser_fingerprint";
import {
  log,
  logError,
  logInfo,
  logDebug,
  logAPICall,
  logAPIResponseWithDetails,
  maskSensitiveData,
  setRequestId,
  getRequestId,
} from "./utils/logger";

/**
 * Browser context configuration for session with optional extension support.
 * 
 * This class provides browser context configuration for cloud sessions and supports
 * automatic extension synchronization when ExtensionOption is provided.
 * 
 * Key Features:
 * - Browser context binding for sessions
 * - Automatic browser data upload on session end
 * - Optional extension integration with automatic context sync generation
 * - Clean API with ExtensionOption encapsulation
 * 
 * Extension Configuration:
 * - **ExtensionOption**: Pass an ExtensionOption object with contextId and extensionIds
 * - **No Extensions**: Don't provide extensionOption parameter (extensionContextSyncs will be undefined)
 * 
 * Usage Examples:
 * ```typescript
 * // With extensions using ExtensionOption
 * import { ExtensionOption } from "./extension";
 * 
 * const extOption = new ExtensionOption(
 *   "my_extensions",
 *   ["ext1", "ext2"]
 * );
 * 
 * const browserContext = new BrowserContext(
 *   "browser_session",
 *   true,
 *   extOption
 * );
 * 
 * // Without extensions (minimal configuration)
 * const browserContext = new BrowserContext(
 *   "browser_session",
 *   true
 * );
 * // extensionContextSyncs will be undefined
 * ```
 */
export class BrowserContext {
  /** ID of the browser context to bind to the session */
  contextId: string;
  /** Whether to automatically upload browser data when the session ends */
  autoUpload: boolean;
  /** Optional browser fingerprint context configuration object containing fingerprintContextId */
  fingerprintContext?: BrowserFingerprintContext;
  /** ID of the fingerprint context for browser fingerprint. Set automatically from fingerprint_context. */
  fingerprintContextId?: string;
  /** Auto-generated context sync for fingerprint. None if no fingerprint configuration provided. */
  fingerprintContextSync?: ContextSync;
  /** Optional extension configuration object containing context_id and extension_ids */
  extensionOption?: ExtensionOption;
  /** ID of the extension context for browser extensions. Set automatically from extension_option. */
  extensionContextId?: string;
  /** List of extension IDs to synchronize. Set automatically from extension_option. */
  extensionIds?: string[];
  /** Auto-generated context syncs for extensions. None if no extension configuration provided. */
  extensionContextSyncs?: ContextSync[];

  /**
   * Initialize BrowserContextImpl with optional extension and fingerprint support.
   *
   * @param contextId - ID of the browser context to bind to the session.
   *                   This identifies the browser instance for the session.
   * @param autoUpload - Whether to automatically upload browser data
   *                    when the session ends. Defaults to true.
   * @param extensionOption - Extension configuration object containing
   *                         contextId and extensionIds. This encapsulates
   *                         all extension-related configuration.
   *                         Defaults to undefined.
   * @param fingerprintContext - Browser fingerprint context configuration object containing
   *                            fingerprintContextId. This encapsulates
   *                            all fingerprint-related configuration.
   *                            Defaults to undefined.
   *
   * Extension Configuration:
   * - **ExtensionOption**: Use extensionOption parameter with an ExtensionOption object
   * - **No Extensions**: Don't provide extensionOption parameter
   *
   * Fingerprint Configuration:
   * - **BrowserFingerprintContext**: Use fingerprintContext parameter with a BrowserFingerprintContext object
   * - **No Fingerprint**: Don't provide fingerprintContext parameter
   *
   * Auto-generation:
   * - extensionContextSyncs is automatically generated when extensionOption is provided
   * - extensionContextSyncs will be undefined if no extensionOption is provided
   * - extensionContextSyncs will be a ContextSync[] if extensionOption is valid
   * - fingerprintContextSync is automatically generated when fingerprintContext is provided
   * - fingerprintContextSync will be undefined if no fingerprintContext is provided
   * - fingerprintContextSync will be a ContextSync if fingerprintContext is valid
   */
  constructor(
    contextId: string,
    autoUpload = true,
    extensionOption?: ExtensionOption,
    fingerprintContext?: BrowserFingerprintContext
  ) {
    this.contextId = contextId;
    this.autoUpload = autoUpload;
    this.extensionOption = extensionOption;
    this.fingerprintContext = fingerprintContext;

    // Handle fingerprint configuration from BrowserFingerprintContext
    if (fingerprintContext) {
      // Extract fingerprint information from BrowserFingerprintContext
      this.fingerprintContextId = fingerprintContext.fingerprintContextId;
      // Auto-generate fingerprint context sync
      this.fingerprintContextSync = this._createFingerprintContextSync();
    } else {
      // No fingerprint configuration provided
      this.fingerprintContextId = undefined;
      this.fingerprintContextSync = undefined;
    }

    // Handle extension configuration from ExtensionOption
    if (extensionOption) {
      // Extract extension information from ExtensionOption
      this.extensionContextId = extensionOption.contextId;
      this.extensionIds = extensionOption.extensionIds;
      // Auto-generate extension context syncs
      this.extensionContextSyncs = this._createExtensionContextSyncs();
    } else {
      // No extension configuration provided
      this.extensionContextId = undefined;
      this.extensionIds = [];
      this.extensionContextSyncs = undefined;
    }
  }

  /**
   * Create ContextSync configurations for browser extensions.
   *
   * This method is called only when extensionOption is provided and contains
   * valid extension configuration (contextId and extensionIds).
   *
   * @returns ContextSync[] - List of context sync configurations for extensions.
   *                         Returns empty list if extension configuration is invalid.
   */
  private _createExtensionContextSyncs(): ContextSync[] {
    if (!this.extensionIds || this.extensionIds.length === 0 || !this.extensionContextId) {
      return [];
    }

    // Create whitelist for each extension ID
    const whiteLists: WhiteList[] = this.extensionIds.map(extId => ({
      path: extId,
      excludePaths: []
    }));

    // Create sync policy for extensions
    const syncPolicy: SyncPolicy = {
      uploadPolicy: {
        ...newUploadPolicy(),
        autoUpload: false
      },
      extractPolicy: {
        ...newExtractPolicy(),
        extract: true,
        deleteSrcFile: true
      },
      deletePolicy: {
        ...newDeletePolicy(),
        syncLocalFile: false
      },
      recyclePolicy: newRecyclePolicy(),
      bwList: {
        whiteLists: whiteLists
      },
      downloadPolicy:{
        autoDownload:true,
        downloadStrategy: DownloadStrategy.DownloadAsync
      },
    };

    // Create context sync for extensions
    const extensionSync = new ContextSync(
      this.extensionContextId,
      "/tmp/extensions/",
      syncPolicy
    );

    return [extensionSync];
  }

  /**
   * Create ContextSync configuration for browser fingerprint.
   *
   * This method is called only when fingerprintContext is provided and contains
   * valid fingerprint configuration (fingerprintContextId).
   *
   * @returns ContextSync - Context sync configuration for fingerprint.
   *                       Returns undefined if fingerprint configuration is invalid.
   */
  private _createFingerprintContextSync(): ContextSync | undefined {
    if (!this.fingerprintContextId || !this.fingerprintContextId.trim()) {
      return undefined;
    }

    // Create sync policy for fingerprint
    const syncPolicy: SyncPolicy = {
      uploadPolicy: {
        ...newUploadPolicy(),
        autoUpload: false
      },
      extractPolicy: {
        ...newExtractPolicy(),
        extract: true,
        deleteSrcFile: true
      },
      deletePolicy: {
        ...newDeletePolicy(),
        syncLocalFile: false
      },
      recyclePolicy: newRecyclePolicy(),
      bwList: {
        whiteLists: []
      }
    };

    // Create context sync for fingerprint
    const fingerprintSync = new ContextSync(
      this.fingerprintContextId,
      BROWSER_FINGERPRINT_PERSIST_PATH,
      syncPolicy
    );

    return fingerprintSync;
  }

  /**
   * Get context syncs for extensions.
   *
   * @returns ContextSync[] - Context sync configurations for extensions. Returns empty list if no extensions configured.
   */
  getExtensionContextSyncs(): ContextSync[] {
    return this.extensionContextSyncs || [];
  }

  /**
   * Get context sync for fingerprint.
   *
   * @returns ContextSync - Context sync configuration for fingerprint.
   *                       Returns undefined if fingerprint configuration is invalid.
   */
  getFingerprintContextSync(): ContextSync | undefined {
    return this.fingerprintContextSync;
  }
  
}

/**
 * Configuration interface for CreateSessionParams
 */
export interface CreateSessionParamsConfig {
   /** Custom labels for the Session. These can be used for organizing and filtering sessions. */
  labels?: Record<string, string>;
  /** Image ID to use for the session. */
  imageId?: string;
  /** List of context synchronization configurations. */
  contextSync?: ContextSync[];
  /** Optional configuration for browser data synchronization. */
  browserContext?: BrowserContext;
  /** Security policy ID (interface field name). */
  policyId?: string;
  /** Beta network ID to bind this session to. */
  betaNetworkId?: string;
  /** Whether to enable browser session recording. */
  enableBrowserReplay?: boolean;
  /** Additional configuration options. */
  extraConfigs?: ExtraConfigs;
  /** Framework identifier for tracking. */
  framework?: string;
  /** Whether to create a VPC-based session. Defaults to false. */
  isVpc?: boolean;
  /** MCP policy id to apply when creating the session (class field name). */
  mcpPolicyId?: string;
}

/**
 * Parameters for creating a session.
 * This interface defines all possible fields that can be used when creating a session.
 */
export interface CreateSessionParamsInterface {
  /** Custom labels for the Session. These can be used for organizing and filtering sessions. */
  labels?: Record<string, string>;
  /** Image ID to use for the session. */
  imageId?: string;
  /** List of context synchronization configurations. */
  contextSync?: ContextSync[];
  /** Optional configuration for browser data synchronization. */
  browserContext?: BrowserContext;
  /** Security policy ID (interface field name). */
  policyId?: string;
  /** Beta network ID to bind this session to. */
  betaNetworkId?: string;
  /** Whether to enable browser session recording. */
  enableBrowserReplay?: boolean;
  /** Additional configuration options. */
  extraConfigs?: ExtraConfigs;
  /** Framework identifier for tracking. */
  framework?: string;
  /** Whether to create a VPC-based session. Defaults to false. */
  isVpc?: boolean;
  /** MCP policy id to apply when creating the session (class field name). */
  mcpPolicyId?: string;
}

/**
 * CreateSessionParams provides a way to configure the parameters for creating a new session
 * in the AgentBay cloud environment.
 */
export class CreateSessionParams implements CreateSessionParamsInterface {
  /** Custom labels for the Session. These can be used for organizing and filtering sessions. */
  public labels: Record<string, string>;

  /** Image ID to use for the session. */
  public imageId?: string;

  /**
   * List of context synchronization configurations.
   * These configurations define how contexts should be synchronized and mounted.
   */
  public contextSync: ContextSync[];

  /** Whether to create a VPC-based session. Defaults to false. */
  public isVpc?: boolean;

  /** MCP policy id to apply when creating the session. */
  public mcpPolicyId?: string;

  /** Security policy ID (interface field name). Maps to mcpPolicyId internally. */
  public policyId?: string;

  /** Beta network ID to bind this session to. */
  public betaNetworkId?: string;

  /** Whether to enable browser session recording. */
  public enableBrowserReplay: boolean;

  /** Additional configuration options. */
  public extraConfigs?: ExtraConfigs;

  /** Framework identifier for tracking. */
  public framework: string;

  public browserContext?: BrowserContext;
  

  constructor(params?: CreateSessionParamsConfig) {
    this.labels = params?.labels || {};
    this.imageId = params?.imageId;
    this.isVpc = params?.isVpc || false;
    // Handle policyId mapping - if policyId is provided, use it, otherwise use mcpPolicyId
    this.mcpPolicyId = params?.policyId || params?.mcpPolicyId;
    this.policyId = params?.policyId || params?.mcpPolicyId;
    this.betaNetworkId = params?.betaNetworkId;
    // Default to true like Python version
    this.enableBrowserReplay = params?.enableBrowserReplay ? params.enableBrowserReplay : true;
    this.extraConfigs = params?.extraConfigs;
    this.framework = params?.framework || "";
    
    // Start with provided contextSync
    let allContextSyncs = params?.contextSync ? [...params.contextSync] : [];

    // Add extension context syncs from browserContext if available
    if (params?.browserContext && params.browserContext.extensionContextSyncs) {
      allContextSyncs = [...allContextSyncs, ...params.browserContext.extensionContextSyncs];
      console.log(`Added ${params.browserContext.extensionContextSyncs.length} extension context sync(s) from BrowserContext`);
    }
    
    // Add fingerprint context sync from browserContext if available
    if (params?.browserContext && params.browserContext.fingerprintContextSync) {
      allContextSyncs = [...allContextSyncs, params.browserContext.fingerprintContextSync];
      console.log("Added fingerprint context sync from BrowserContext");
    }

    this.contextSync = allContextSyncs;
    this.browserContext = params?.browserContext;
  }
}